decode.go

```diff
@@ -308,7 +308,7 @@ func (md *MetaData) unifySlice(data interface{}, rv reflect.Value) error {
 		return badtype("slice", data)
 	}
 	sliceLen := datav.Len()
-	if rv.IsNil() {
+	if rv.IsNil() || rv.Len() != datav.Len() {
```

Comment:
Instead you might want to only do this if `rv.Cap() < datav.Len()` and otherwise `rv.SetLen(datav.Len())`. But it's an optimization that probably doesn't matter for most TOML use case.